internal/pkg/agent/storage/store/state_store.go

```diff
@@ -104,136 +122,63 @@ func NewStateStore(log *logger.Logger, store storeLoad) (*StateStore, error) {
 	if err != nil {
-		return nil, err
-	}
-
-	state := stateT{
-		ackToken: sr.AckToken,
-		queue:    sr.Queue,
+		return nil, fmt.Errorf("could not parse store content: %w", err)
 	}
 
-	if sr.Action != nil {
-		if sr.Action.IsDetected != nil {
-			state.action = &fleetapi.ActionUnenroll{
-				ActionID:   sr.Action.ID,
-				ActionType: sr.Action.Type,
-				IsDetected: *sr.Action.IsDetected,
-			}
-		} else {
-			state.action = &fleetapi.ActionPolicyChange{
-				ActionID:   sr.Action.ID,
-				ActionType: sr.Action.Type,
-				Policy:     conv.YAMLMapToJSONMap(sr.Action.Policy), // Fix Policy, in order to make it consistent with the policy received from the fleet gateway as nested map[string]interface{}
-			}
-		}
+	if st.Version != Version {
+		return nil, fmt.Errorf(
+			"invalid state store version, got %q isntead of %s",
+			st.Version, Version)
 	}
 
 	return &StateStore{
 		log:   log,
 		store: store,
-		state: state,
+		state: st,
 	}, nil
 }
 
-func migrateStateStore(ctx context.Context, log *logger.Logger, actionStorePath, stateStorePath string, storageOpts ...storage.EncryptedOptionFunc) (err error) {
-	log = log.Named("state_migration")
-	actionDiskStore, err := storage.NewDiskStore(actionStorePath)
-	if err != nil {
-		return fmt.Errorf("error creating disk store: %w", err)
-	}
-
-	stateDiskStore, err := storage.NewEncryptedDiskStore(ctx, stateStorePath, storageOpts...)
-	if err != nil {
-		return fmt.Errorf("error instantiating encrypted disk store: %w", err)
-	}
-
-	stateStoreExits, err := stateDiskStore.Exists()
-	if err != nil {
-		log.Errorf("failed to check if state store %s exists: %v", stateStorePath, err)
-		return err
-	}
-
-	// do not migrate if the state store already exists
-	if stateStoreExits {
-		log.Debugf("state store %s already exists", stateStorePath)
-		return nil
-	}
-
-	actionStoreExits, err := actionDiskStore.Exists()
-	if err != nil {
-		log.Errorf("failed to check if action store %s exists: %v", actionStorePath, err)
-		return err
-	}
-
-	// delete the actions store file upon successful migration
-	defer func() {
-		if err == nil && actionStoreExits {
-			err = actionDiskStore.Delete()
-			if err != nil {
-				log.Errorf("failed to delete action store %s exists: %v", actionStorePath, err)
-			}
-		}
-	}()
-
-	// nothing to migrate if the action store doesn't exists
-	if !actionStoreExits {
-		log.Debugf("action store %s doesn't exists, nothing to migrate", actionStorePath)
-		return nil
-	}
+// readState parsed the content from reader as JSON to state.
+// It's mostly to abstract the parsing of the date so different functions can
+// reuse this.
+func readState(reader io.ReadCloser) (state, error) {
+	st := state{}
 
-	actionStore, err := newActionStore(log, actionDiskStore)
+	data, err := io.ReadAll(reader)
 	if err != nil {
-		log.Errorf("failed to create action store %s: %v", actionStorePath, err)
-		return err
+		return state{}, fmt.Errorf("could not read store state: %w", err)
```

Comment:
Thanks yeah I don't think I followed it correctly the first time. What was tripping me up is [this](https://github.com/elastic/elastic-agent/blob/bb44a11df7a638f84f806357388b65c3c0a6b213/internal/pkg/agent/storage/store/state_store.go#L113-L116):

```go
// NewStateStore creates a new state store.
func NewStateStore(log *logger.Logger, store saveLoader) (*StateStore, error) {
	// If the store exists we will read it, if an error is returned we log it
	// and return an empty store.
	reader, err := store.Load()
```

Specifically that the error handling here is to return an empty store, but further down we don't follow this approach and return an error:

````go
	st, err := readState(reader)
	if err != nil {
		return nil, fmt.Errorf("could not parse store content: %w", err)
	}

	if st.Version != Version {
		return nil, fmt.Errorf(
			"invalid state store version, got %q isntead of %s",
			st.Version, Version)
	}
```

The error handling isn't consistent. Going off the comment, I was thinking the empty `state{}` returned by `readState` would be returned with an unset Version field. 

This makes me wonder if we should be return an empty store on error in all cases, and not just the call to `store.Load()`. Is there a reason for the error handling to be different here?

Reply:
It was like that already, this is the current behaviour so I did not change it. What I remember from my tests before I was on sick leave, the agent would run normally but with an empty policy if the store is empty because something deleted/messed up with the store.